### DISCUSSION TITLE
Batch Export PDFs from Zotero

### DISCUSSION THREAD

**Discussion by abdullahahmed-dev** [windingwind/zotero-actions-tags#379] [Action Scripts]
### Description

Select multiple files with attached PDFs in Zotero, then run this script to automatically export all attached PDFs to your downloads folder . This convenient tool saves you time and effort, allowing you to quickly access and organize your PDF files."

#### Usage
In the script, first configure the path your downloads folder path (or your folder of choice) in this line. 
```JS
var downloadsPath = '/Users/user/Downloads';
```

Then trigger the script and it will create a new folder in your Downloads Folder with all your selected PDF files. 

#### Preview
![Export PDF Script](https://github.com/user-attachments/assets/5cb18b5b-1956-4c91-a026-2ada86647947)

### Event

None

### Operation

Script

### Data

```javascript
/**
 * Batch export PDFs from Zotero
 * @author quantumfrost00
 * @usage Configure the path to your foler in var downloadsPath = '/Users/user/Downloads'
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/379
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/
 */

(async function() {
  var zp = Zotero.getActiveZoteroPane();
  var items = zp.getSelectedItems();

  if (!items.length) {
    zp.displayAlert(null, "Please select some items first.");
    return;
  }

  var downloadsPath = '/Users/user/Downloads';
  var lockFile = Zotero.File.pathToFile(downloadsPath);
  lockFile.append('.zotero_export_lock');

  // Check if an export is already in progress
  if (lockFile.exists()) {
    zp.displayAlert(null, "An export is already in progress. Please wait for it to complete.");
    return;
  }

  // Create lock file
  lockFile.create(Components.interfaces.nsIFile.NORMAL_FILE_TYPE, 0o666);

  try {
    var folderName = 'Zotero_Export_' + new Date().toISOString().replace(/[-:]/g, '').split('.')[0];
    var exportDir = Zotero.File.pathToFile(downloadsPath);
    exportDir.append(folderName);

    // If the folder already exists, use it instead of creating a new one
    if (!exportDir.exists()) {
      exportDir.create(Components.interfaces.nsIFile.DIRECTORY_TYPE, 0o755);
    }

    var exportCount = 0;

    for (let item of items) {
      let attachments = item.getAttachments(false);
      for (let attID of attachments) {
        let att = Zotero.Items.get(attID);
        if (att.attachmentContentType === 'application/pdf') {
          let file = att.getFile();
          if (file && file.exists()) {
            // Check if file already exists in export directory
            let destFile = exportDir.clone();
            destFile.append(file.leafName);
            if (!destFile.exists()) {
              file.copyTo(exportDir, file.leafName);
              exportCount++;
            }
          }
        }
      }
    }

    if (exportCount > 0) {
      zp.displayAlert(null, `Exported ${exportCount} PDFs to ${exportDir.path}`);
    } else {
      zp.displayAlert(null, "No new PDFs found to export.");
    }
  } finally {
    // Always remove the lock file, even if an error occurs
    if (lockFile.exists()) {
      lockFile.remove(false);
    }
  }
})();
```

**Comment by windingwind** [comment]
https://github.com/windingwind/zotero-actions-tags/blob/master/README.md#custom-script

> Please always use the `items` instead of `ZoteroPane.getSelectedItems()`. The action can be triggered from entrances outside the library, where the items are not from `ZoteroPane`.

**Reply by abdullahahmed-dev** [reply]
Sorry but I am new to this, I couldn't get it to work with the change you suggested

**Comment by mvmalyi** [comment]
Hi @quantumfrost00 ! I saw you had a batch tag operation command in there, too. Could you please share your script for that one? I've managed to set up a command for batch adding a pre-difined tag within the script, but I couldn't find a way to select the tag when calling the command, or for removing the tags set on items. I'm curious what your solution is!

**Reply by abdullahahmed-dev** [reply]
Hi @mvmalyi it is actually an excellent community script by @thalliant-ai. You can find it [here](https://github.com/windingwind/zotero-actions-tags/discussions/351). When triggered it comes up with multiple options for tags and works great.

**Reply by mvmalyi** [reply]
Thanks a lot for sharing! Your script for PDF export is also super helpful.

**Reply by abdullahahmed-dev** [reply]
Thankyou

**Comment by ihateforking** [comment]
Is anyone looking for a complete OpenAI Agent based solution?! I’ve fully tested and set up a working end-to-end implementation based on it.

### 1. Install the Actions & Tags Plugin
Download the latest .xpi from the [releases page](https://github.com/windingwind/zotero-actions-tags/releases).
In Zotero, go to Tools → Add-ons, click the gear icon → Install Add-on From File…
Select the .xpi you downloaded and restart Zotero.

### 2. Prepare the Export Script
Replace /path/to/your/Downloads with your own folder path.

> Windows: C:\\Users\\YourName\\Downloads
> macOS: /Users/YourName/Downloads
> Linux: /home/yourname/Downloads

```
/**
 * Batch export PDFs from Zotero (custom Actions & Tags script)
 * Set downloadsPath to the folder where you want the exported PDFs.
 */
(async function() {
  const Zotero = require('Zotero');
  const zp = Zotero.getActiveZoteroPane();

  // If the script is triggered for a single item (variable 'item' is set), skip to avoid duplicate runs.
  if (typeof item !== 'undefined' && item !== null) {
    return;
  }

  // Use the 'items' array supplied by the plugin.
  if (!items || items.length === 0) {
    zp.displayAlert(null, "Please select some items first.");
    return;
  }

  // Adjust this path to your system
  const downloadsPath = '/path/to/your/Downloads';

  const exportBase = Zotero.File.pathToFile(downloadsPath);
  const lockFile = exportBase.clone();
  lockFile.append('.zotero_export_lock');

  // Avoid running multiple exports at once
  if (lockFile.exists()) {
    zp.displayAlert(null, "An export is already in progress. Please wait for it to complete.");
    return;
  }

  // Create lock file
  lockFile.create(Components.interfaces.nsIFile.NORMAL_FILE_TYPE, 0o666);

  try {
    const folderName = 'Zotero_Export_' + new Date().toISOString().replace(/[-:]/g, '').split('.')[0];
    const exportDir = exportBase.clone();
    exportDir.append(folderName);
    if (!exportDir.exists()) {
      exportDir.create(Components.interfaces.nsIFile.DIRECTORY_TYPE, 0o755);
    }

    let exportCount = 0;

    for (const it of items) {
      const attachments = it.getAttachments(false);
      for (const attID of attachments) {
        const att = Zotero.Items.get(attID);
        if (att.attachmentContentType === 'application/pdf') {
          const file = att.getFile();
          if (file && file.exists()) {
            const destFile = exportDir.clone();
            destFile.append(file.leafName);
            if (!destFile.exists()) {
              file.copyTo(exportDir, file.leafName);
              exportCount++;
            }
          }
        }
      }
    }

    if (exportCount > 0) {
      zp.displayAlert(null, `Exported ${exportCount} PDFs to ${exportDir.path}`);
    } else {
      zp.displayAlert(null, "No new PDFs found to export.");
    }
  } finally {
    // Remove the lock file so another export can run
    if (lockFile.exists()) {
      lockFile.remove(false);
    }
  }
})();
```
### 3. Add the Script as a Custom Actio

1. Open Zotero settings→ Actions & Tags tab.
2. Click the “+” button to add a new action.
3. Set Operation to Script.
4. Paste the script above into the Data field.
5. Enter a menu label (e.g., Batch Export PDFs).
6. (Optional) Assign a keyboard shortcut.
7. Click Save.

### 4. Run the Script
In your Zotero library, select the items you want to export PDFs from.
Right-click and choose Batch Export PDFs from the context menu (or use your shortcut).
PDFs will be copied into a new, time-stamped folder inside your chosen directory.

### Notes
Each export run creates a new folder named like Zotero_Export_20250804T135910.
If nothing is exported, make sure your selected items have PDF attachments.
If you get an error, double-check the downloadsPath and that the folder exists.
Script uses a lock file (.zotero_export_lock) to avoid duplicate runs.

You’re ready!
You can now batch-export PDFs from your Zotero library using the Actions & Tags plugin.

**Reply by LXSKING** [reply]
可用，好用，支持！

**Comment by dosidicus-gigas** [comment]
Hey! Perhaps I came too late, but I added the script as a custom action with the plugin "Actions & Tags", as you described. I select the articles with the PDF attachments, copy the script, change the download path, run the script and...nothing happens! I don't even know if it started or not; no error message, nothing. Any idea why it is not working?

**Reply by dosidicus-gigas** [reply]
I fixed the problem, it had to do with the "zp.displayAlert()" function, I replaced it with the "alert()" instead anddd it works ^^

**Comment by dsmabulage** [comment]
if using windows client ` var downloadsPath = 'C:\\Users\\<user name>\\Documents\\Tests';`
